File: objectModel/TypeScript/Persistence/CdmFolder/PurposeReferencePersistence.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import {
    CdmCorpusContext,
    cdmObjectType,
    CdmPurposeDefinition,
    CdmPurposeReference,
    CdmTraitGroupReference,
    CdmTraitReference,
    CdmTraitReferenceBase
} from '../../internal';
import { CdmFolder } from '..';
import { cdmObjectRefPersistence } from './cdmObjectRefPersistence';
import {
    PurposeReference
} from './types';
import * as utils from './utils';

export class PurposeReferencePersistence extends cdmObjectRefPersistence {
    public static fromData(ctx: CdmCorpusContext, object: string | PurposeReference): CdmPurposeReference {
        if (!object) {
            return;
        }

        let simpleReference: boolean = true;
        let optional: boolean | undefined;
        let appliedTraits: (CdmTraitReference | CdmTraitGroupReference)[];
        let purpose: string | CdmPurposeDefinition;
        
        if (typeof (object) === 'string') {
            purpose = object;
        } else {
            simpleReference = false;
            optional = object.optional;

            if (typeof (object.purposeReference) === 'string') {
                purpose = object.purposeReference;
            } else {
                purpose = CdmFolder.PurposePersistence.fromData(ctx, object.purposeReference);
            }
        }

        const purposeReference: CdmPurposeReference = ctx.corpus.MakeRef(cdmObjectType.purposeRef, purpose, simpleReference);

        if (optional !== undefined) {
            purposeReference.optional = optional;
        }

        if (typeof (object) !== 'string') {
            appliedTraits = utils.createTraitReferenceArray(ctx, object.appliedTraits);
        }
        utils.addArrayToCdmCollection<CdmTraitReferenceBase>(purposeReference.appliedTraits, appliedTraits);

        return purposeReference;
    }
}

File: objectModel/Python/cdm/persistence/cdmfolder/purpose_reference_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Union, Optional, List

from cdm.enums import CdmObjectType
from cdm.objectmodel import CdmCorpusContext, CdmPurposeReference

from . import utils
from .cdm_object_ref_persistence import CdmObjectRefPersistence
from .purpose_persistence import PurposePersistence
from .types import PurposeReference


class PurposeReferencePersistence(CdmObjectRefPersistence):
    @staticmethod
    def from_data(ctx: CdmCorpusContext, data: Union[str, PurposeReference]) -> Optional[CdmPurposeReference]:
        if not data:
            return None

        simple_reference = True
        applied_traits = None
        purpose = None

        if isinstance(data, str):
            purpose = data
        else:
            simple_reference = False
            if isinstance(data.purposeReference, str):
                purpose = data.purposeReference
            else:
                purpose = PurposePersistence.from_data(ctx, data.purposeReference)

        purpose_reference = ctx.corpus.make_ref(CdmObjectType.PURPOSE_REF, purpose, simple_reference)

        if not isinstance(data, str):
            applied_traits = utils.create_trait_reference_array(ctx, data.get('appliedTraits'))
            purpose_reference.applied_traits.extend(applied_traits)

        return purpose_reference
